pkg/acsengine/transform.go

```diff
@@ -197,62 +199,117 @@ func removeImageReference(logger *logrus.Entry, resourceProperties map[string]in
+		if !ok {
+			logger.Warnf("Template improperly formatted for field name: %s", resourcesFieldName)
+			continue
+		}
+
+		resourceType, ok := resourceMap[typeFieldName].(string)
+		if !ok {
+			continue
+		}
+
+		if !(resourceType == vmResourceType || resourceType == vmExtensionType) {
+			continue
+		}
 
-	for _, computeResourceType := range computeResourceTypes {
-		resources := templateMap[resourcesFieldName].([]interface{})
+		resourceName, ok := resourceMap[nameFieldName].(string)
+		if !ok {
+			logger.Warnf("Template improperly formatted for field name: %s", nameFieldName)
+			continue
+		}
 
-		agentPoolIndex := -1
-		//remove agent nodes resources
-		for index, resource := range resources {
-			resourceMap, ok := resource.(map[string]interface{})
+		if strings.EqualFold(resourceType, vmResourceType) &&
+			strings.Contains(resourceName, "variables('masterVMNamePrefix')") {
+			resourceProperties, ok := resourceMap[propertiesFieldName].(map[string]interface{})
 			if !ok {
-				logger.Warnf("Template improperly formatted")
+				logger.Warnf("Template improperly formatted for field name: %s, resource name: %s", propertiesFieldName, resourceName)
 				continue
 			}
 
-			resourceType, ok := resourceMap[typeFieldName].(string)
-			if !ok || resourceType != computeResourceType {
+			storageProfile, ok := resourceProperties[storageProfileFieldName].(map[string]interface{})
+			if !ok {
+				logger.Warnf("Template improperly formatted: %s", storageProfileFieldName)
 				continue
 			}
 
-			resourceName, ok := resourceMap[nameFieldName].(string)
-			if !ok {
-				logger.Warnf("Template improperly formatted")
+			dataDisks := storageProfile[dataDisksFieldName].([]interface{})
+			dataDisk, ok := dataDisks[0].(map[string]interface{})
+			dataDisk[createOptionFieldName] = "attach"
+		}
+
+		tags, ok := resourceMap[tagsFieldName].(map[string]interface{})
+		poolName := fmt.Sprint(tags["poolName"]) // poolName tag exists on agents only
+
+		if resourceType == vmResourceType {
+			logger.Infoln(fmt.Sprintf("Evaluating if agent pool: %s, resource: %s needs to be removed", poolName, resourceName))
+			// Not an agent (could be a master VM)
+			if tags["poolName"] == nil || strings.Contains(resourceName, "variables('masterVMNamePrefix')") {
 				continue
 			}
 
-			if strings.EqualFold(resourceType, vmResourceType) &&
-				strings.Contains(resourceName, "variables('masterVMNamePrefix')") {
-				resourceProperties, ok := resourceMap[propertiesFieldName].(map[string]interface{})
-				if !ok {
-					logger.Warnf("Template improperly formatted")
-					continue
-				}
+			logger.Infoln(fmt.Sprintf("agentPoolsToPreserve: %v...", agentPoolsToPreserve))
 
-				storageProfile, ok := resourceProperties[storageProfileFieldName].(map[string]interface{})
-				if !ok {
-					logger.Warnf("Template improperly formatted")
-					continue
+			if agentPoolsToPreserve == nil || len(agentPoolsToPreserve) == 0 || agentPoolsToPreserve[poolName] != true {
```

Comment:
I am pulling the pool name out of the VM name and comparing it with the tag that was created during template creation so they should be the same unless someone modifies it. I can write a custom comparator for this to ignore case when searching a map but don't see an immediate issue for now.